src/java/org/jivesoftware/openfire/net/SASLAuthentication.java

```diff
@@ -192,6 +192,24 @@ public static Element getSASLMechanismsElement( ClientSession session )
     {
         final Element result = DocumentHelper.createElement( new QName( "mechanisms", new Namespace( "", SASL_NAMESPACE ) ) );
         for (String mech : getSupportedMechanisms()) {
+            if (mech.equals("EXTERNAL")) {
+                boolean trustedCert = false;
+                if (session.isSecure()) {
+                    final LocalClientSession localClientSession = (LocalClientSession)session;
+                    if (localClientSession != null) {
+                        final Connection connection = localClientSession.getConnection();
+                        final KeyStore keyStore = connection.getConfiguration().getIdentityStore().getStore();
+                        final KeyStore trustStore = connection.getConfiguration().getTrustStore().getStore();
+                        final X509Certificate trusted = CertificateManager.getEndEntityCertificate(connection.getPeerCertificates(), keyStore, trustStore);
+                        if (trusted != null) {
+                            trustedCert = true;
```

Comment:
Why not 'continue' here? Perhaps some debug logging - but otherwise, looks good to me.